vllm/entrypoints/openai/tools.py

```diff
@@ -0,0 +1,291 @@
+                    instructions = tool_params.function_guided + "\n" + self.render_tool(
+                        tool, tool_params=tool_params) + "\n"
+                    instructions += tool_params.function_call_instruct
+                    return instructions.strip()
+            return ""  # Tool not found. What should we do ?
+        else:
+            instructions = tool_params.function_list_start + "\n"
+            for tool in tools_list:
+                instructions += self.render_tool(
+                    tool, tool_params=tool_params) + "\n"
+            instructions = instructions.strip()
+            instructions += "\n" + tool_params.function_list_end + "\n"
+            instructions += tool_params.function_call_instruct
+            return instructions.strip()
+
+
+class OpenAIToolsPrompter:
+    """
+    https://platform.openai.com/docs/assistants/tools
+    """
+
+    def __init__(self, debug: bool):
+        self.debug = debug
+        self.template = ToolsCallsTemplate()
+
+    def content_from_assistant(self, message: ChatCompletionAssistantMessage,
+                               tool_params: VllmToolsTemplate) -> str:
+        text = self.template.render_toolcalls(message.tool_calls,
+                                              tool_params=tool_params)
+        if message.content is None or len(message.content.strip()) == 0:
+            return text
+        else:
+            return message.content.strip() + "\n\n" + text
+
+    def content_from_tool(self, message: ChatCompletionToolMessage,
+                          tool_params: VllmToolsTemplate) -> str:
+        return self.template.render_toolresponse(message,
+                                                 tool_params=tool_params)
+
+    def inject_prompt(self, request: ChatCompletionRequest):
+        """ Generate and inject the prompt for tools calls. """
+        if request.tools is not None and len(request.tools):
+            if (isinstance(request.tool_choice,
+                           ChatCompletionNamedToolChoiceParam)):
+                if request.tool_choice.type == "function":
+                    select_tool_choice = request.tool_choice.function.name
+                else:
+                    select_tool_choice = None
+            else:
+                select_tool_choice = None
+            text_inject = self.template.render_toolslist(
+                tool_choice=select_tool_choice,
+                tools_list=request.tools,
+                tool_params=request.tool_params)
+            if isinstance(request.messages, str):
+                request.messages = request.messages.strip(
+                ) + "\n\n" + text_inject
+            elif isinstance(request.messages,
+                            List) and len(request.messages) >= 1:
+                request.messages[0].content = request.messages[
+                    0].content.strip() + "\n\n" + text_inject
```

Comment:
This just adds the system prompt to the first message. In almost every case tool calls should be a system prompt message (except for chat formats that don't do system prompts, but that's a marginal amount), so instead of checking for the first message and just adding the tool call prompt, it might be better to do something like this:

```python
if isinstance(request.messages, str):
    # suggested change: for string prompts, put the tool usage before the prompt 
    request.messages = text_inject + '\n\n' + request.messages.strip() 
elif isinstance(request.messages, List) and len(request.messages) >= 1:
    # suggested change: check to see if the first message is a system message. If so, edit it.
    # otherwise, add a system prompt
    if request.messages[0].role == 'system':
        request.messages[0].content = request.messages[0].content.strip() + '\n\n' + text_inject
    else:
        request.messages = [{'role': 'system', 'content': text_inject}] + request.messages
```